google_takeout_to_sqlite/utils.py

```diff
@@ -53,3 +58,117 @@ def id_for_location_history(row):
         datetime.datetime.utcfromtimestamp(int(row["timestampMs"]) / 1000).isoformat(),
         first_six,
     )
+
+
+def get_mbox(mbox_file):
+    num_errors = 0
+    print("Preparing to process emails...")
+    mbox = mailbox.mbox(mbox_file)
+    print("Processing {} emails".format(len(mbox)))
+
+    # These are all the Gmail email fields available
+    # ['X-GM-THRID', 'X-Gmail-Labels', 'Delivered-To', 'Received', 'Received',
+    # 'Return-Path', 'Received', 'Received-SPF', 'Authentication-Results',
+    # 'Received', 'Mailing-List', 'Precedence', 'List-Post', 'List-Help',
+    # 'List-Unsubscribe', 'List-Subscribe', 'Delivered-To', 'Received',
+    # 'Message-ID', 'Date', 'From', 'To', 'MIME-Version', 'Content-Type',
+    # 'Content-Transfer-Encoding', 'X-Nabble-From', 'X-pstn-neptune',
+    # 'X-pstn-levels', 'X-pstn-settings', 'X-pstn-addresses', 'Subject']
+
+    for email in track(mbox):
+        try:
+            message = {}
+            message["Message-Id"] = email["Message-Id"]
+            message["X-GM-THRID"] = email["X-GM-THRID"]
+            message["X-Gmail-Labels"] = email["X-Gmail-Labels"]
+
+            # These following try/excepts are here because for some reason
+            # these items returned from the mbox module are sometimes strings
+            # and sometimes headers and sometimes None.
+
+            try:
+                email["From"].decode("utf-8")
+            except AttributeError:
+                message["From"] = str(email["From"])
+            try:
+                email["To"].decode("utf-8")
+            except AttributeError:
+                message["To"] = str(email["To"])
+
+            try:
+                email["Subject"].decode("utf-8")
+            except AttributeError:
+                message["Subject"] = str(email["Subject"])
+
+            message["date"] = get_message_date(email.get("Date"), email.get_from())
+            message["body"] = get_email_body(email)
+
+            yield message
+        except (TypeError, ValueError, AttributeError, LookupError) as e:
+            # How does this project want to handle logging? For now we're just
+            # printing out variables
+            num_errors = num_errors + 1
+            print("Errors: {}".format(num_errors))
+            print(traceback.format_exc())
+            continue
+
+
+def save_emails(db, mbox_file):
+    """
+    Import Gmail mbox from google takeout
+    """
+    db["mbox_emails"].upsert_all(
```

Comment:
A fix for the problem I had where my `body` column ended up being a `BLOB` rather than text would be to explicitly create the table first.

You can do that like so:
```python
db["mbox_emails"].create({
    "id": str,
    "X-GM-THRID": str,
    "X-Gmail-Labels": str,
    "From": str,
    "To": str,
    "Subject": str,
    "when": str,
    "body": str,
}, pk="id")
```